File: 45. Jump Game II/solution.py
class Solution:
    def jump(self, nums: list[int]) -> int:
        dp = [0 for i in range(len(nums))]
        for i in range(len(nums)):
            if dp[i] or i == 0:
                for j in range(1, nums[i] + 1):
                    if i + j >= len(nums):
                        break
                    if dp[i + j]:
                        dp[i + j] = min(dp[i] + 1, dp[i + j])
                    else:
                        dp[i + j] = dp[i] + 1
        return dp[-1]

File: 45. Jump Game II/solution.go
package main

func jump(nums []int) int {
	if len(nums) == 1 {
		return 0
	}
	preJump := 0
	currJump := nums[0]
	times := 1
	for currJump < len(nums)-1 {
		tmpJump := preJump
		for i := preJump; i <= currJump; i++ {
			if nums[i]+i > tmpJump {
				tmpJump = nums[i] + i
			}
		}
		preJump = currJump
		currJump = tmpJump
		times++
	}
	return times
}

func main() {
	nums := []int{2, 3, 1, 1, 4}
	println(jump(nums))
}
